File: backend/app/ai/retrieval.py
"""Knowledge base retrieval."""

from __future__ import annotations

import math
import re
from collections import Counter
from collections.abc import Iterable, Sequence
from dataclasses import dataclass
from typing import Optional

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..config import settings
from ..models import Article
from .providers import get_provider

_WORD_RE = re.compile(r"[a-z0-9']+")

_STOPWORDS = frozenset(
    """a an and are as at be been but by can cant could did do does doesnt for from
    get got had has have how i im in is it its many me much my no not of on or our
    so than that the their them then there these they this to too very was we were
    what when where which who why will with would you your""".split()
)

_COVERAGE_MIDPOINT = 2.2

_MIN_TERM_WEIGHT = 1.2

_LEXICAL_WEIGHT = 0.6
_SEMANTIC_WEIGHT = 0.4


_VERB_SUFFIXES = ("ingly", "edly", "ing", "ied", "ed", "ly")

_STEM_FLOOR = 5

_ES_PLURAL_ENDINGS = ("ses", "xes", "zes", "ches", "shes")


def stem(word: str) -> str:
    """Reduce a word to a rough root, so tenses and plurals match."""
    if len(word) < _STEM_FLOOR:
        return word

    if word.endswith("ies") and len(word) > 4:
        return word[:-3] + "y"
    if word.endswith(_ES_PLURAL_ENDINGS):
        word = word[:-2]
    elif word.endswith("s") and not word.endswith("ss"):
        word = word[:-1]

    if len(word) < _STEM_FLOOR:
        return word

    for suffix in _VERB_SUFFIXES:
        if word.endswith(suffix) and len(word) - len(suffix) >= 3:
            root = word[: -len(suffix)]
            # A doubled consonant from "shipped" or "planning" collapses.
            if len(root) > 3 and root[-1] == root[-2] and root[-1] not in "aeiou":
                root = root[:-1]
            return root

    if len(word) >= _STEM_FLOOR and word.endswith("e"):
        return word[:-1]
    return word


def tokenize(text: str) -> list[str]:
    """Words worth matching on, lowercased, stopped and stemmed."""
    return [
        stem(word)
        for word in _WORD_RE.findall(text.lower())
        if word not in _STOPWORDS
    ]


@dataclass(slots=True)
class Hit:
    article: Article
    score: float
    coverage: float = 0.0

    def excerpt(self, query: str, length: int = 240) -> str:
        """A window of the body around the first query term, else the opening."""
        body = " ".join(self.article.body.split())
        if not body:
            return self.article.summary or ""

        lowered = body.lower()
        position = -1
        for term in tokenize(query):
            position = lowered.find(term)
            if position != -1:
                break

        if position == -1 or position < length:
            start = 0
        else:
            start = max(0, position - length // 3)
        return _window(body, start, length)


_SENTENCE_LOOKBACK = 160


def _window(body: str, start: int, length: int) -> str:
    """Cut a readable window out of a body of text."""
    if start > 0:
        sentence = _sentence_start(body, start)
        if sentence is not None:
            start = sentence
        else:
            space = body.find(" ", start)
            start = start if space == -1 else space + 1

    whole_sentence = False
    end = start + length
    if end < len(body):
        sentence = _sentence_end(body, start, end)
        if sentence is not None:
            end = sentence
            whole_sentence = True
        else:
            space = body.rfind(" ", start, end)
            end = end if space <= start else space

    prefix = "..." if start > 0 else ""
    suffix = "..." if end < len(body) and not whole_sentence else ""
    return f"{prefix}{body[start:end].strip()}{suffix}"


def _sentence_start(body: str, position: int) -> int | None:
    """Index just after the nearest sentence end before `position`, if close."""
    floor = max(0, position - _SENTENCE_LOOKBACK)
    best = max(body.rfind(f"{mark} ", floor, position) for mark in ".!?")
    return best + 2 if best != -1 else None


def _sentence_end(body: str, start: int, position: int) -> int | None:
    """Index just after the last sentence that finishes inside the window."""
    floor = max(start, position - _SENTENCE_LOOKBACK)
    best = max(body.rfind(f"{mark} ", floor, position) for mark in ".!?")
    if best == -1:
        if position >= len(body) and body[position - 1 : position] in ".!?":
            return position
        return None
    return best + 1


def _document_text(article: Article) -> str:
    """Title and summary repeat so matches there outrank body only matches."""
    return " ".join(
        [
            article.title,
            article.title,
            article.summary or "",
            " ".join(article.tags or []),
            article.category or "",
            article.body,
        ]
    )


def bm25_scores(query: str, documents: Sequence[str], k1: float = 1.5, b: float = 0.75) -> list[float]:
    """Standard BM25 over an in memory corpus."""
    terms = tokenize(query)
    if not terms or not documents:
        return [0.0] * len(documents)

    tokenized = [tokenize(doc) for doc in documents]
    lengths = [len(doc) or 1 for doc in tokenized]
    avg_length = sum(lengths) / len(lengths)
    frequencies = [Counter(doc) for doc in tokenized]

    document_frequency = Counter()
    for counts in frequencies:
        for term in set(terms):
            if counts[term]:
                document_frequency[term] += 1

    total = len(documents)
    scores = [0.0] * total
    for index, counts in enumerate(frequencies):
        score = 0.0
        for term in terms:
            frequency = counts[term]
            if not frequency:
                continue
            n_q = document_frequency[term]
            idf = math.log(1 + (total - n_q + 0.5) / (n_q + 0.5))
            norm = frequency * (k1 + 1)
            denom = frequency + k1 * (1 - b + b * lengths[index] / avg_length)
            score += idf * norm / denom
        scores[index] = score
    return scores


def cosine(a: Sequence[float], b: Sequence[float]) -> float:
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b, strict=True))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(y * y for y in b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot / (norm_a * norm_b)


def _normalise(values: Iterable[float]) -> list[float]:
    values = list(values)
    peak = max(values, default=0.0)
    if peak <= 0:
        return [0.0] * len(values)
    return [value / peak for value in values]


async def embed_query(query: str) -> Optional[list[float]]:
    if not settings.EMBEDDING_MODEL:
        return None
    try:
        vectors = await get_provider().embed([query])
    except Exception:  # noqa: BLE001, retrieval must degrade, never fail hard
        return None
    return vectors[0] if vectors else None


async def embed_article(article: Article) -> Optional[list[float]]:
    """Compute and attach an article's embedding, if a model is configured."""
    if not settings.EMBEDDING_MODEL:
        return None
    text = f"{article.title}\n\n{article.summary or ''}\n\n{article.body}"[:8000]
    try:
        vectors = await get_provider().embed([text])
    except Exception:  # noqa: BLE001, indexing does not have to succeed
        return None
    return vectors[0] if vectors else None


async def search(
    db: AsyncSession,
    workspace_id: str,
    query: str,
    *,
    limit: int = 5,
    include_internal: bool = True,
    min_score: float = 0.05,
    min_coverage: float = 0.15,
) -> list[Hit]:
    """Rank published articles for a query. Returns at most `limit` hits."""
    query = (query or "").strip()
    if not query:
        return []

    statement = select(Article).where(
        Article.workspace_id == workspace_id,
        Article.status == "published",
    )
    if not include_internal:
        statement = statement.where(Article.visibility == "public")

    articles = list((await db.execute(statement)).scalars())
    if not articles:
        return []

    documents = [_document_text(article) for article in articles]
    lexical = _normalise(bm25_scores(query, documents))

    semantic = [0.0] * len(articles)
    vector = await embed_query(query)
    if vector:
        semantic = _normalise(
            cosine(vector, article.embedding or []) if article.embedding else 0.0
            for article in articles
        )
        blended = [
            _LEXICAL_WEIGHT * lex + _SEMANTIC_WEIGHT * sem
            for lex, sem in zip(lexical, semantic, strict=True)
        ]
    else:
        blended = lexical

    coverage = _coverage_scores(query, documents)
    hits = [
        Hit(article=article, score=round(score, 4), coverage=cover)
        for article, score, cover in zip(articles, blended, coverage, strict=True)
        if score >= min_score
    ]
    hits = [hit for hit in hits if hit.coverage >= min_coverage]
    hits.sort(key=lambda hit: hit.score, reverse=True)
    return hits[:limit]


def _coverage_scores(query: str, documents: Sequence[str]) -> list[float]:
    """How much of the question each document actually covers, from 0 to 1."""
    terms = list(dict.fromkeys(tokenize(query)))
    if not terms or not documents:
        return [0.0] * len(documents)

    present = [set(tokenize(doc)) for doc in documents]
    total = len(documents)
    weights = {}
    for term in terms:
        seen = sum(1 for words in present if term in words)
        # Same shape as the BM25 idf, so the two agree on what is informative.
        idf = math.log(1 + (total - seen + 0.5) / (seen + 0.5))
        weights[term] = max(idf, _MIN_TERM_WEIGHT)

    scores = []
    for words in present:
        matched = sum(weights[term] for term in terms if term in words)
        scores.append(round(matched / (matched + _COVERAGE_MIDPOINT), 4))
    return scores

File: src/lib/demo/retrieval.ts
/** Lexical retrieval for demo mode. */

import type { Article } from '../types';

const WORD = /[a-z0-9']+/g;

const STOPWORDS = new Set(
  `a an and are as at be been but by can cant could did do does doesnt for from
   get got had has have how i im in is it its me my no not of on or our so than
   that the their them then there these they this to too very was we were what
   when where which who why will with would you your many much`
    .split(/\s+/)
    .filter(Boolean),
);

const VERB_SUFFIXES = ['ingly', 'edly', 'ing', 'ied', 'ed', 'ly'];
const ES_PLURAL_ENDINGS = ['ses', 'xes', 'zes', 'ches', 'shes'];

/** Below this a word is too short to strip safely. */
const STEM_FLOOR = 5;

/** Reduce a word to a rough root, so tenses and plurals match. */
export function stem(word: string): string {
  if (word.length < STEM_FLOOR) return word;

  if (word.endsWith('ies') && word.length > 4) return `${word.slice(0, -3)}y`;
  let base = word;
  if (ES_PLURAL_ENDINGS.some((ending) => base.endsWith(ending))) {
    base = base.slice(0, -2);
  } else if (base.endsWith('s') && !base.endsWith('ss')) {
    base = base.slice(0, -1);
  }
  if (base.length < STEM_FLOOR) return base;

  for (const suffix of VERB_SUFFIXES) {
    if (base.endsWith(suffix) && base.length - suffix.length >= 3) {
      let root = base.slice(0, -suffix.length);
      // A doubled consonant from shipped or planning collapses.
      if (root.length > 3 && root.at(-1) === root.at(-2) && !'aeiou'.includes(root.at(-1)!)) {
        root = root.slice(0, -1);
      }
      return root;
    }
  }

  // A trailing silent e goes last, so arrive and arrived both end at arriv.
  if (base.length >= STEM_FLOOR && base.endsWith('e')) return base.slice(0, -1);
  return base;
}

export function tokenize(text: string): string[] {
  return (text.toLowerCase().match(WORD) ?? [])
    .filter((word) => !STOPWORDS.has(word))
    .map(stem);
}

function documentText(article: Article): string {
  // Title counted twice, so a match there outranks one in the body alone.
  return [
    article.title,
    article.title,
    article.summary ?? '',
    (article.tags ?? []).join(' '),
    article.category,
    article.body,
  ].join(' ');
}

export function bm25(query: string, documents: string[], k1 = 1.5, b = 0.75): number[] {
  const terms = tokenize(query);
  if (terms.length === 0 || documents.length === 0) return documents.map(() => 0);

  const tokenized = documents.map(tokenize);
  const lengths = tokenized.map((doc) => doc.length || 1);
  const avgLength = lengths.reduce((sum, value) => sum + value, 0) / lengths.length;

  const frequencies = tokenized.map((doc) => {
    const counts = new Map<string, number>();
    for (const word of doc) counts.set(word, (counts.get(word) ?? 0) + 1);
    return counts;
  });

  const unique = [...new Set(terms)];
  const documentFrequency = new Map<string, number>();
  for (const term of unique) {
    documentFrequency.set(term, frequencies.filter((counts) => counts.has(term)).length);
  }

  const total = documents.length;
  return frequencies.map((counts, index) => {
    let score = 0;
    for (const term of terms) {
      const frequency = counts.get(term) ?? 0;
      if (!frequency) continue;
      const nq = documentFrequency.get(term) ?? 0;
      const idf = Math.log(1 + (total - nq + 0.5) / (nq + 0.5));
      const norm = frequency * (k1 + 1);
      const denom = frequency + k1 * (1 - b + (b * lengths[index]) / avgLength);
      score += (idf * norm) / denom;
    }
    return score;
  });
}

export interface Hit {
  article: Article;
  /** Rank within this result set, normalised so the best hit is always 1. */
  score: number;
  /** Absolute match quality, comparable across searches. See coverageScores. */
  coverage: number;
}

/** Matched rarity at which coverage reaches one half. */
const COVERAGE_MIDPOINT = 2.2;

/** The least a matched content word can be worth, for small knowledge bases. */
const MIN_TERM_WEIGHT = 1.2;

/** How much of the question each document actually answers, from 0 to 1. */
function coverageScores(query: string, documents: string[]): number[] {
  const terms = [...new Set(tokenize(query))];
  if (terms.length === 0 || documents.length === 0) return documents.map(() => 0);

  const present = documents.map((doc) => new Set(tokenize(doc)));
  const total = documents.length;
  const weights = new Map<string, number>();
  for (const term of terms) {
    const seen = present.filter((words) => words.has(term)).length;
    const idf = Math.log(1 + (total - seen + 0.5) / (seen + 0.5));
    // Floored, because idf collapses towards zero on a small corpus.
    weights.set(term, Math.max(idf, MIN_TERM_WEIGHT));
  }

  return present.map((words) => {
    let matched = 0;
    for (const term of terms) if (words.has(term)) matched += weights.get(term) ?? 0;
    return Number((matched / (matched + COVERAGE_MIDPOINT)).toFixed(4));
  });
}

export function searchArticles(
  articles: Article[],
  query: string,
  { limit = 5, includeInternal = true, minScore = 0.05, minCoverage = 0.15 } = {},
): Hit[] {
  const trimmed = query.trim();
  if (!trimmed) return [];

  const pool = articles.filter(
    (article) =>
      article.status === 'published' && (includeInternal || article.visibility === 'public'),
  );
  if (pool.length === 0) return [];

  const documents = pool.map(documentText);
  const raw = bm25(trimmed, documents);
  const peak = Math.max(...raw, 0);
  if (peak <= 0) return [];

  const coverage = coverageScores(trimmed, documents);
  return pool
    .map((article, index) => ({
      article,
      score: Number((raw[index] / peak).toFixed(4)),
      coverage: coverage[index],
    }))
    .filter((hit) => hit.score >= minScore && hit.coverage >= minCoverage)
    .sort((a, b) => b.score - a.score)
    .slice(0, limit);
}

/** A window of the article body around the first matching term. */
export function excerpt(article: Article, query: string, length = 240): string {
  const body = article.body.split(/\s+/).join(' ');
  if (!body) return article.summary ?? '';

  const lowered = body.toLowerCase();
  let position = -1;
  for (const term of tokenize(query)) {
    position = lowered.indexOf(term);
    if (position !== -1) break;
  }

  // A match inside the first window reads best from the top of the article,
  // rather than shaving a word or two off the opening sentence to centre it.
  const from = position === -1 || position < length
    ? 0
    : Math.max(0, position - Math.floor(length / 3));
  return windowAround(body, from, length);
}

/** How far back to look for a sentence start before cutting on a word instead. */
const SENTENCE_LOOKBACK = 160;

/** Cut a readable window out of a body of text. */
function windowAround(body: string, from: number, length: number): string {
  let start = from;
  if (start > 0) {
    const sentence = sentenceStart(body, start);
    if (sentence !== null) {
      start = sentence;
    } else {
      const space = body.indexOf(' ', start);
      start = space === -1 ? start : space + 1;
    }
  }

  let whole = false;
  let end = start + length;
  if (end < body.length) {
    const sentence = sentenceEnd(body, start, end);
    if (sentence !== null) {
      end = sentence;
      whole = true;
    } else {
      const space = body.lastIndexOf(' ', end - 1);
      end = space <= start ? end : space;
    }
  }

  const prefix = start > 0 ? '...' : '';
  const suffix = end < body.length && !whole ? '...' : '';
  return `${prefix}${body.slice(start, end).trim()}${suffix}`;
}

/** Index just after the nearest sentence end before `position`, when close. */
function sentenceStart(body: string, position: number): number | null {
  const floor = Math.max(0, position - SENTENCE_LOOKBACK);
  const window = body.slice(floor, position);
  const best = Math.max(...['. ', '! ', '? '].map((mark) => window.lastIndexOf(mark)));
  return best === -1 ? null : floor + best + 2;
}

/** Index just after the last sentence that finishes inside the window. */
function sentenceEnd(body: string, start: number, position: number): number | null {
  const floor = Math.max(start, position - SENTENCE_LOOKBACK);
  const window = body.slice(floor, position);
  const best = Math.max(...['. ', '! ', '? '].map((mark) => window.lastIndexOf(mark)));
  if (best === -1) {
    if (position >= body.length && ['.', '!', '?'].includes(body.slice(position - 1, position))) {
      return position;
    }
    return null;
  }
  return floor + best + 1;
}
